### PR TITLE
#30 Fix/sync bioprinter

### PR DESCRIPTION
**Pull request by zuev93**

### CODE REVIEW

File: MultiplayerMod/Game/UI/Screens/ImmigrantScreenPatch.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                SetDeliverablesData(__instance);
+            }
+        )
+    );
+
+    public static async Task<List<ITelepadDeliverable>> WaitForAllDeliverablesReady(ImmigrantScreen instance) {
+        var currentDelay = 0;
+        while (currentDelay < maxWaitMS) {
+            var readyDeliverables = instance.containers?.Select(
+                a => a switch {
```
Comment (polycone) [inline MultiplayerMod/Game/UI/Screens/ImmigrantScreenPatch.cs:43]: What is `a`?

File: MultiplayerMod/Game/UI/Screens/Events/ImmigrantScreenEvents.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    public static event Action<ITelepadDeliverable> Proceed;
+
+    [HarmonyPostfix]
+    [HarmonyPatch(nameof(ImmigrantScreen.Initialize))]
+    private static void InitializePatch(ImmigrantScreen __instance) => PatchControl.RunIfEnabled(
+        () => new TaskFactory(Container.Get<UnityTaskScheduler>()).StartNew(
+            async () => {
+                if (ImmigrantScreenPatch.Deliverables != null) return;
+
+                var readyDeliverables = await ImmigrantScreenPatch.WaitForAllDeliverablesReady(__instance);
```
Comment (polycone) [inline MultiplayerMod/Game/UI/Screens/Events/ImmigrantScreenEvents.cs:26]: Maybe this should be extracted from `ImmigrantScreenPatch`? Just to be consistent with responsibility and refrain patch methods calling from another patch.

File: MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/CarePackageInstanceDataSurrogate.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+public class CarePackageInstanceDataSurrogate : ISerializationSurrogate, ISurrogateType {
+
+    public Type Type => typeof(CarePackageContainer.CarePackageInstanceData);
+
+    public void GetObjectData(object obj, SerializationInfo info, StreamingContext context) {
+        var node = (CarePackageContainer.CarePackageInstanceData) obj;
+        info.AddValue("facadeID", node.facadeID);
+        info.AddValue("info.id", node.info.id);
+        info.AddValue("info.quantity", node.info.quantity);
+        info.AddValue("info.facadeID", node.info.facadeID);
```
Comment (polycone) [inline MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/CarePackageInstanceDataSurrogate.cs:15]: Is this different with `facadeId`?

File: MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/CarePackageInstanceDataSurrogate.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+using System.Runtime.Serialization;
+
+namespace MultiplayerMod.Platform.Steam.Network.Messaging.Surrogates;
+
+public class CarePackageInstanceDataSurrogate : ISerializationSurrogate, ISurrogateType {
+
+    public Type Type => typeof(CarePackageContainer.CarePackageInstanceData);
+
+    public void GetObjectData(object obj, SerializationInfo info, StreamingContext context) {
+        var node = (CarePackageContainer.CarePackageInstanceData) obj;
```
Comment (polycone) [inline MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/CarePackageInstanceDataSurrogate.cs:11]: Hmm, why is it a `node`? I see a "care package container" and "instance data", are graphs involved somehow?

File: MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/MinionStartingStatsSurrogate.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        info.AddValue("NameStringKey", node.NameStringKey);
+        info.AddValue("GenderStringKey", node.GenderStringKey);
+        info.AddValue("stressTrait", node.stressTrait?.Id);
+        info.AddValue("joyTrait", node.joyTrait?.Id);
+        info.AddValue("congenitaltrait", node.congenitaltrait?.Id);
+        info.AddValue("stickerType", node.stickerType);
+        info.AddValue("voiceIdx", node.voiceIdx);
+        info.AddValue("personality", node.personality?.Id);
+        info.AddValue("Traits", node.Traits.Select(a => a?.Id).ToArray());
+        info.AddValue("StartingLevels.Keys", node.StartingLevels.Keys.ToArray());
```
Comment (polycone) [inline MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/MinionStartingStatsSurrogate.cs:24]: What type is `node.StartingLevels`? Can't we serialize it as a raw dictionary🤔?

**Review by polycone — APPROVED** [review]

File: MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/MinionStartingStatsSurrogate.cs (hunk 5 above)
Comment (zuev93) [reply MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/MinionStartingStatsSurrogate.cs:24]: Nope. Dictionary won't get serialised. It seems like here we need to add primitive types only

Comment (polycone) [reply MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/MinionStartingStatsSurrogate.cs:24]: Wow, okay 🫤

File: MultiplayerMod/Game/UI/Screens/ImmigrantScreenPatch.cs (hunk 1 above)
Comment (zuev93) [reply MultiplayerMod/Game/UI/Screens/ImmigrantScreenPatch.cs:43]: Done

File: MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/CarePackageInstanceDataSurrogate.cs (hunk 4 above)
Comment (zuev93) [reply MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/CarePackageInstanceDataSurrogate.cs:11]: Done

File: MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/CarePackageInstanceDataSurrogate.cs (hunk 3 above)
Comment (zuev93) [reply MultiplayerMod/Platform/Steam/Network/Messaging/Surrogates/CarePackageInstanceDataSurrogate.cs:15]: They might be different if info.facadeId is 'SELECTRANDOM' than facadeID will be choosen upon generation and must be sent with generated value instead if `SELECTRANDOM`

File: MultiplayerMod/Game/UI/Screens/Events/ImmigrantScreenEvents.cs (hunk 2 above)
Comment (zuev93) [reply MultiplayerMod/Game/UI/Screens/Events/ImmigrantScreenEvents.cs:26]: Done, moved this to ScreenUtils

**Pull request merged**